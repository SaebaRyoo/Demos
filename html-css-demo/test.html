
<!DOCTYPE html>
<html>
    <body>
        <div id='demo'>
            <ol>
                <li>test</li>
            </ol>
        </div>
    </body>
    <script type="text/javascript">
      var compareVersion = function(version1, version2) {
          var v1Arr = version1.split('.')
          var v2Arr = version2.split('.')

          var maxLen = Math.max(v1Arr.length, v2Arr.length)
          var idx = 0;
          while (idx < maxLen) {
              var v1Item = parseInt(v1Arr[idx]) || 0
              var v2Item = parseInt(v2Arr[idx]) || 0



              if (v1Item > v2Item ) {
                  return 1
              }

              if (v1Item < v2Item) {
                  return -1
              }

              idx++
          }
          return 0
      };
      const result = compareVersion('3', '2.7.070')
      console.log(result)
    </script>
</html>
