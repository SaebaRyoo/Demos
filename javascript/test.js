// Example test:   ([3, 1, 2], [2, 3, 1])
// OK

// Example test:   ([1, 2, 1], [2, 3, 3])
// OK

// Example test:   ([1, 2, 3, 4], [2, 1, 4, 4])
// WRONG ANSWER (got True expected False)

// Example test:   ([1, 2, 3, 4], [2, 1, 4, 3])
// WRONG ANSWER (got True expected False)

// Example test:   ([1, 2, 2, 3, 3], [2, 3, 3, 4, 5])

function solution(A, B) {
  const n = A.length;
  const graph = new Map(); // Create an empty graph

  // Build the graph
  for (let i = 0; i < n; i++) {
    const from = A[i];
    const to = B[i];
    if (!graph.has(from)) {
      graph.set(from, new Set());
    }
    if (!graph.has(to)) {
      graph.set(to, new Set());
    }
    graph.get(from).add(to); // Add the edge to the "from" vertex's adjacency set
  }

  const visited = new Set();
  const stack = new Set();

  const dfs = (node) => {
    visited.add(node);
    stack.add(node);

    const neighbors = graph.get(node);
    for (const neighbor of neighbors) {
      if (stack.has(neighbor)) {
        // Found a cycle
        return true;
      }
      if (!visited.has(neighbor)) {
        // Continue the DFS if not visited yet
        if (dfs(neighbor)) {
          return true;
        }
      }
    }

    stack.delete(node);
    return false;
  };

  // Try DFS from every unvisited node
  for (const [node] of graph) {
    if (!visited.has(node)) {
      if (dfs(node)) {
        return true;
      }
    }
  }

  return false;
}
