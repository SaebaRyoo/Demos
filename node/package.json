{
  "name": "node",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "dependencies": {
    "puppeteer": "^19.4.1"
  },
  "devDependencies": {
    "babel-preset-env": "^1.7.0"
  }
}
