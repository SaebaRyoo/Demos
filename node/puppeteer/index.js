const puppeteer = require('puppeteer');

(async () => {
  const browser = await puppeteer.launch({headless: false});
  const page = await browser.newPage();

  await page.goto('https://etherscan.io/address/0x0264438173ebcd19651e6b2a5f63dcec0f0afc4d#code');

  // Wait for the results page to load and display the results.
  const divIdCodeSelector = '#dividcode';
  const divIdCodeHandle = await page.waitForSelector(divIdCodeSelector);

  // Extract the results from the page.
  const result = await page.evaluate(async (divCode) => {

    // find title which contains 'Contract Source Code'
    const contractSourceCodeTitleEle = [...document.querySelectorAll('#dividcode h4.card-header-title')].filter(ele => {
      return ele.innerHTML.match(/contract source code/gi)
    })[0];


    // ele.parentElement.querySelector('a.js-clipboard').click()
    // const code = await navigator.clipboard.readText()
    const contractSourceCodeWrapper = contractSourceCodeTitleEle.parentElement.parentElement;
    // 根据Contract Source Code 找到父节点的父节点，然后在这个结构找查找.texttext-secondary
    const sourceCodeNames =
      [...contractSourceCodeWrapper.querySelectorAll('span.text-secondary')]
      .filter((_, index) => index > 0) // 过滤掉第一个'span.text-secondary',它是标题中的元素，无用
      .map((ele) => {
        // find code
        ele.parentElement.querySelector('a.js-clipboard').click();
        navigator.clipboard.readText().then((clipText) => document.querySelector("body").innerText = clipText)
        return {
          filname: ele.innerHTML,
        }
      });

    await sourceCodeNames;


    // final result
    return {
      // sourceCode: sourceCodeNames.map((name) => ({filename: name})),
      sourceCode: sourceCodeNames,
      // code
    }
  }, divIdCodeHandle);

  // Print all the files.
  console.log(result);

  // await browser.close();
})();
