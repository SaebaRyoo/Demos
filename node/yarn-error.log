Arguments: 
  /Users/zhuqingyuan/.nvm/versions/node/v14.19.1/bin/node /usr/local/Cellar/yarn/1.22.19/libexec/bin/yarn.js add puppeter

PATH: 
  /Users/zhuqingyuan/.nvm/versions/node/v14.19.1/bin:/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/Users/zhuqingyuan/.nvm/versions/node/v14.19.1/bin:/usr/local/mysql/bin:/usr/local/mysql/support-files:/usr/local/sbin:/Users/zhuqingyuan/Documents/maven/apache-maven-3.6.3/bin:/usr/local/redis-6.2.4/src:/usr/local/mysql/bin:/usr/local/mysql/support-files:/usr/local/sbin:/Users/zhuqingyuan/Documents/maven/apache-maven-3.6.3/bin:/usr/local/redis-6.2.4/src

Yarn version: 
  1.22.19

Node version: 
  14.19.1

Platform: 
  darwin x64

Trace: 
  Error: https://registry.npm.taobao.org/puppeter: [NOT_FOUND] puppeter not found
      at Request.params.callback [as _callback] (/usr/local/Cellar/yarn/1.22.19/libexec/lib/cli.js:66145:18)
      at Request.self.callback (/usr/local/Cellar/yarn/1.22.19/libexec/lib/cli.js:140890:22)
      at Request.emit (events.js:400:28)
      at Request.<anonymous> (/usr/local/Cellar/yarn/1.22.19/libexec/lib/cli.js:141862:10)
      at Request.emit (events.js:400:28)
      at Gunzip.<anonymous> (/usr/local/Cellar/yarn/1.22.19/libexec/lib/cli.js:141784:12)
      at Object.onceWrapper (events.js:519:28)
      at Gunzip.emit (events.js:400:28)
      at endReadableNT (internal/streams/readable.js:1334:12)
      at processTicksAndRejections (internal/process/task_queues.js:82:21)

npm manifest: 
  {
    "name": "node",
    "version": "1.0.0",
    "main": "index.js",
    "license": "MIT"
  }

yarn manifest: 
  No manifest

Lockfile: 
  No lockfile
