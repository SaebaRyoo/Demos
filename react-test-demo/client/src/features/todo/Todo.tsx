import React, { useEffect, useState } from 'react';
import { List, Checkbox, Input, Button } from 'antd';
import classes from "./Todo.module.css"
import { setInputValue, createTodo, fetchTodos, updateTodo, deleteTodoById } from './Todo.slice'
import { useAppSelector, useAppDispatch } from '../../hooks/typedHooks';

const Todo: React.FC = () => {

  const inputValue = useAppSelector((state) => state.todo.inputValue);
  const data = useAppSelector((state) => state.todo.data);
  const dispatch = useAppDispatch();


  useEffect(() => {
    dispatch(fetchTodos())
  }, [])

  const handleChange = (checked: boolean, todo: API.TodoItem) => {
    dispatch(updateTodo({id: todo.id, name: todo.name, done: checked}))
  }

  return (
    <div className={classes.container} aria-label='test-todo'>
      <Input
        placeholder="请输入todo"
        aria-label='create-todo'
        value={inputValue}
        onChange={(e) => dispatch(setInputValue(e.target.value))}
        onPressEnter={(e) => {
          if(!inputValue) return;
          dispatch(createTodo())
        }}
      />
      <List
        itemLayout="horizontal"
        split={false}
        dataSource={data}
        renderItem={(item, index) => (
          <List.Item className={item.done ?'done' : ''}>
            <Checkbox aria-label={`todo-checkbox-${index}`} checked={item.done} onChange={(e) => handleChange(e.target.checked, item)} />
            <span>{item.name}</span>
            <Button onClick={() => {dispatch(deleteTodoById(item.id))}} >删除</Button>
          </List.Item>
        )}
      />
    </div>
  )
};

export default Todo;
