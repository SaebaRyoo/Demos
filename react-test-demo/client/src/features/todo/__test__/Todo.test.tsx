import { screen, act } from '@testing-library/react';
import React from 'react';
import Todo from '../Todo'
import classes from '../TODO.module.css'
import {renderWithProviders} from '../../../test/testUils'
import { setInputValue } from '../Todo.slice'
import { setupServer } from 'msw/node'
import { todo } from '../../../__mocks__/api/todoMock';

const server = setupServer(...todo)

// 在所有测试开始之前启用 API mock
beforeAll(() => server.listen())

// 在每个测试之后关闭 API mock
afterEach(() => server.resetHandlers())

// 所有测试结束时关闭 API mock
afterAll(() => server.close())

describe('<Todo/> ', () => {
  beforeAll(() => {
    // Error: Uncaught [TypeError: window.matchMedia is not a function]
    // The above error occurred in the <List> component:
    //       at List (react-test-demo/client/node_modules/antd/lib/list/index.js:36:7)
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  })

  it(`it should correctly with default value `, async () => {
    const preloadState = {
      todo: {
        inputValue: 'test',
        data: [
          {
              id: '1',
              name: 'hello',
              done: false,
          }
        ]
      }
    };

    const {container} = renderWithProviders(<Todo/>, {preloadedState: preloadState})
    expect(container).toMatchSnapshot()
  })

  it(`it should fetch mock data and render it correctly`, async () => {

    const { user } = renderWithProviders(<Todo/>)
    expect(await screen.findByText('To learn node')).toBeInTheDocument()

    const input = await screen.findByRole('textbox', {name: 'create-todo'});
    await user.type(input, 'learning react test')
    expect(input).toHaveValue('learning react test')
    expect(input).not.toHaveValue('0')
    expect(input).not.toHaveValue('1')

    // mock user operation
    input.focus();
    await user.keyboard('{enter/}')
    const test = await screen.findByText('learning react test')
    expect(test).toBeInTheDocument()

  })

  it(`it should change the status of todo`, async () => {
    const { user } = renderWithProviders(<Todo/>)
    const testDom = await screen.findByRole("checkbox", {name: 'todo-checkbox-0'})
    expect(testDom).not.toBeChecked()
    testDom.focus();
    await user.click(testDom)
    expect(testDom).toBeChecked()
  })

  it(`Todo should be deleted when button was clicked`, async () => {
    const { user } = renderWithProviders(<Todo/>)
    expect(await screen.findByText('To learn node')).toBeInTheDocument()
    const delBtn0 = await screen.findByRole('button', {name: 'todo-btn-0'})
    await user.click(delBtn0)
    expect(await screen.findByText('No data')).toBeInTheDocument()
  })
})
