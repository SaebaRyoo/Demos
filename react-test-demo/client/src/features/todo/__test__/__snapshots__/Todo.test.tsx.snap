// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo/>  it should correctly with default value  1`] = `
<div>
  <div
    aria-label="test-todo"
    class="container"
  >
    <input
      aria-label="create-todo"
      class="ant-input css-dev-only-do-not-override-k83k30"
      placeholder="请输入todo"
      type="text"
      value="test"
    />
    <div
      class="ant-list css-dev-only-do-not-override-k83k30"
    >
      <div
        class="ant-spin-nested-loading css-dev-only-do-not-override-k83k30"
      >
        <div
          class="ant-spin-container"
        >
          <ul
            class="ant-list-items"
          >
            <li
              class="ant-list-item"
            >
              <label
                class="ant-checkbox-wrapper css-dev-only-do-not-override-k83k30"
              >
                <span
                  class="ant-checkbox css-dev-only-do-not-override-k83k30"
                >
                  <input
                    aria-label="todo-checkbox-0"
                    class="ant-checkbox-input"
                    type="checkbox"
                    value=""
                  />
                  <span
                    class="ant-checkbox-inner"
                  />
                </span>
              </label>
              <span>
                hello
              </span>
              <button
                aria-label="todo-btn-0"
                class="ant-btn css-dev-only-do-not-override-k83k30 ant-btn-default"
                type="button"
              >
                <span>
                  删 除
                </span>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
